## Introduction:

Write a function called Pumpkin that uses functions from thinkstats.py to compute the mean, variance and standard deviation of the pumpkins weights in the previous section (in the textbook). 

There is no space to utilize thinkstats.py code .. hence, all the calculations will be implemented using Julia's built in functions and the double check will be made through calculate them again using excel .. 

## Pumpkins example in details .. 

Suppose I grow several varieties in my garden, and one day I harvest three decorative pumpkins that are 1 pound each, two pie pumpkins that are 3 pounds each, and one Atlantic Giant® pumpkin that weighs 591 pounds. The mean of this sample is 100 pounds.

In [1]:
pumpkins_sizes = [1, 1, 1, 3, 3, 591]

6-element Array{Int64,1}:
   1
   1
   1
   3
   3
 591

In [2]:
m = mean(pumpkins_sizes) ## similar to what has been mentioned

100.0

In [3]:
var(pumpkins_sizes) ## similar as excel, differs from python 

57860.4

Differs from python results of 48217.0 .. 

Let's try to write it ourselves .. 

In [4]:
v = 0
for pumpkin in pumpkins_sizes
    v = (sum(pumpkin - m)^2)/(length(pumpkins_sizes)-1) ## sample variance
end

In [5]:
v

48216.2

This is closer to python result .. 


Hence, standard devaition is expected to be closer as well ..  

In [6]:
std_ = sqrt(v)

219.58187539047935

Very close to python results of (219.58369702689677) .. as expected .. 

## However .. 
applying the std built in function is expected to vary from the above result (like what has been done in the variance value) 

In [7]:
## in the 1st case (built in std) .. 

std(pumpkins_sizes) ## similar as excel, differs from python

240.54188824402289

This is far as well from python results of (219.58369702689677)

## Justification .. 

Perhanps coding the varaince calculations from scratch results in a nearly similar results like python .. however, I tried to calculate the variance for the above data using excel and the output was (57860.4) which is exactly the same as julia built in function ! :S  


## Conclusion:
In python ==> 
    
    built in function, funtion from textbook implementation ==> matches 
    
In Julia ==> 
    - built in function, python ==> doesn't match 
    - built in function, excel ==> matches 
    - end to end impelemenation, python ==> matches 
    
    
### Trials to understand the mechanism on ver() in Julia: 
after searching the internet, many refrences have been found that describe the mechanism of var() in Julia .. 

i.e. Page 11 here ==> https://media.readthedocs.org/pdf/statsbasejl/latest/statsbasejl.pdf

There are many versions of var() .. and they all basicaly calculate weighted variance .. 

In [12]:
var(pumpkins_sizes, mean = 100) ## calculate weighted variance 

57860.4

In [11]:
var(pumpkins_sizes, mean = 0) ## far from all the values 

69860.4

In [18]:
var(pumpkins_sizes, mean = 1) ## expected ! .. as the mean of 1 is near 3 values in the data .. 

69621.6

In [17]:
var(pumpkins_sizes, mean = 500) ## expected ! 

249860.4

as long as mean value increases, varaince increases as well .. which makes sense cause mean misrepresent more data points .. 

However, I think calculating var value from scratch is more rebust cause it deosn't depend on weighted mean, and has justifiable result ..  